parallel-crypto3/libs/parallel-zk/include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp

```diff
@@ -468,14 +469,30 @@ namespace nil {
                     }
 
                     std::unique_ptr<std::vector<polynomial_dfs_type>> prepare_lookup_value(
-                            const polynomial_dfs_type& mask_assignment) {
+                        const polynomial_dfs_type &mask_assignment,
+                        const polynomial_dfs_type &lagrange0
+                    ) {
                         PROFILE_SCOPE("Lookup argument preparing lookup value");
 
                         // Prepare lookup value
                         auto lookup_value_ptr = std::make_unique<std::vector<polynomial_dfs_type>>();
                         for (std::size_t t_id = 0; t_id < lookup_tables.size(); t_id++) {
                             const plonk_lookup_table<FieldType> &l_table = lookup_tables[t_id];
-                            const polynomial_dfs_type &lookup_tag = plonk_columns.selector(l_table.tag_index);
+                            polynomial_dfs_type lookup_tag;
+                            if( l_table.tag_index >=0 )
+                                lookup_tag = plonk_columns.selector(l_table.tag_index);
+                            else{
+                                switch( l_table.tag_index ){
+                                case PLONK_SPECIAL_SELECTOR_ALL_USABLE_ROWS_SELECTED:
+                                    lookup_tag = mask_assignment;
+                                    break;
+                                case PLONK_SPECIAL_SELECTOR_ALL_NON_FIRST_USABLE_ROWS_SELECTED:
+                                    lookup_tag = mask_assignment - lagrange0;
+                                    break;
+                                default:
+                                    BOOST_ASSERT_MSG(false, "Unknown special selector");
```

Comment:
Maybe move the general case into the "default"? lookup_tag = plonk_columns.selector(l_table.tag_index);

Reply:
Replaced switch by direct if-elses